Add sync indicator

paulsonnentag opened the pull request:
Uses the new sync state API to show a sync indicator:

- if the current document is in sync with the server, a circle with a solid line is shown
- if the current document is out of sync with the server, a circle with a dashed line is shown


Currently, this is hardcoded to report the sync state of the sync server peer ID. Eventually, we should extract the hook logic and move it to automerge-repo-react-hooks.

To test this branch you need to checkout automerge-repo on the `peer-sync-report-paul` branch and link `automerge-repo`   in TEE.

geoffreylitt commented on src/components/App.tsx:
```diff
@@ -1,3 +1,4 @@
+import * as A from "@automerge/automerge"
```
should be importing from `next`?

geoffreylitt replied:
per pvh it doesn't matter for the methods we're using. but as a general practice in TEE i think it's good because then we don't accidentally use old APIs in other places where we deal w text

pvh commented on src/components/SyncIndicator.tsx:
```diff
@@ -0,0 +1,121 @@
+  )
+}
+
+const SYNC_SERVER_PEER_ID = "storage-server-sync-automerge-org" as PeerId
```
be nice to get this from the peers list instead

geoffreylitt commented on src/components/SyncIndicator.tsx:
```diff
@@ -0,0 +1,121 @@
+
+  useEffect(() => {
+    const onOnline = () => {
+      console.log("online")
```
remove console.log statements

geoffreylitt commented:
@pvh and I did some cleanup work

- made it use new official version of repo
- temporarily switched it to work with a local sync server to make it easier to test various states (need to revert this part before merge!)
- added a quick draft of new icons

I think it's looking pretty good, generally seems to work as expected. 

Remaining work: add debounce logic + avoid showing sync error unless it really seems like an error (sorry, I'm in a rush writing this so I can't say precisely what the remaining work is but hopefully @paulsonnentag if you play w/ it should be pretty clear what's needed to make it feel good in practice)

paulsonnentag commented:
- I've adjusted the spacing of the icon so it doesn't shift around when going through the different states. 
- There are two error case that I'm handling

  - not connected to the sync server while online
    - initially we wait for a bit until the connection to the sync server is established. If we don't have a connection after 2 seconds we go to the error state
  - pending changes while online and the server hasn't responded in a while (3 seconds)

- To test the error states I've switched between using the remote server and the local server
   - The first error mostly occurs with the remote server
   - To test the second error case I've modified `node_modules/@automerge/automerge-repo-network-websocket/dist/NodeWServerAdapter.js` so it doesn't process messages after a timeout   
   

   ```javascript
   
    const startup = Date.now()

    socket.on("message", message => {        
      if ((Date.now() - startup) < 5000) {
        console.log("sync", message)
        this.receiveMessage(message, socket)
      } else {
       console.log("...and it's broken")              
      }
    });
     ```

geoffreylitt commented on src/components/SyncIndicator.tsx:
```diff
@@ -0,0 +1,191 @@
+const INITIAL_CONNECTION_WAIT = 2000
+
+export const SyncIndicator = ({ handle }: { handle: DocHandle<unknown> }) => {
+  // todo: sync shouldn't flicker. It should be based on if we are making progress syncing the changes.
```
@paulsonnentag I think this todo comment is out of date and can be removed, yes?

geoffreylitt commented on src/components/SyncIndicator.tsx:
```diff
@@ -0,0 +1,191 @@
+interface TimeoutConfig {
+  duration: number
+  timestamp?: number,
+  isActive?: boolean
```
does this basically mean "we're actively checking to see if there's been a timeout"? could you add comments to the fields on this interface

geoffreylitt commented on src/components/SyncIndicator.tsx:
```diff
@@ -0,0 +1,191 @@
+  }
+}
+
+export function arraysEqual(a, b) {
```
why is this exported? want to move it into utils?

geoffreylitt commented:
I added a hover popover with some basic information. Open to feedback on how to present / communicate this info but it seems like a good start to me.

<img width="358" alt="CleanShot 2023-11-14 at 10 26 44@2x" src="https://github.com/inkandswitch/tiny-essay-editor/assets/934016/9564ac45-dcba-4d0f-b835-26946ebbb221">

<img width="389" alt="CleanShot 2023-11-14 at 10 27 24@2x" src="https://github.com/inkandswitch/tiny-essay-editor/assets/934016/e17e54cd-3289-4a3d-8dbc-e381abbc98f3">

<img width="412" alt="CleanShot 2023-11-14 at 10 28 00@2x" src="https://github.com/inkandswitch/tiny-essay-editor/assets/934016/abea42a0-dbed-4010-9a45-34d6733e2564">

<img width="354" alt="CleanShot 2023-11-14 at 10 29 44@2x" src="https://github.com/inkandswitch/tiny-essay-editor/assets/934016/566e3f18-1786-4328-871f-443261f70721">

@paulsonnentag can you get rid of the flickering error state when you start typing after a long pause? I think we need to track the time of the last change and compare the last sync against the last change?

geoffreylitt commented:
I revised the popover design a bit

<img width="1790" alt="CleanShot 2023-11-14 at 10 47 58@2x" src="https://github.com/inkandswitch/tiny-essay-editor/assets/934016/a33e9501-1610-4ea1-a6ba-6d35eacd5008">
